junit-jupiter-params/src/main/java/org/junit/jupiter/params/provider/CsvArgumentsProvider.java

```diff
@@ -54,47 +72,86 @@ public void accept(CsvSource annotation) {
+		return Arrays.stream(this.annotation.value())
+				.map(line -> parseLine(index.getAndIncrement(), line))
 				.map(Arguments::of);
 		// @formatter:on
 	}
 
-	private String[] parseLine(long index, String line) {
-		String[] parsedLine = null;
+	private List<String[]> parseTextBlock(String textBlock) {
 		try {
-			parsedLine = this.csvParser.parseLine(line + LINE_SEPARATOR);
-			if (parsedLine != null && !this.nullValues.isEmpty()) {
-				for (int i = 0; i < parsedLine.length; i++) {
-					if (this.nullValues.contains(parsedLine[i])) {
-						parsedLine[i] = null;
-					}
-				}
+			String sanitizedTextBlock = sanitizeTextBlock(textBlock);
+			AtomicInteger index = new AtomicInteger(1);
+			List<String[]> parsedLines = this.csvParser.parseAll(new StringReader(sanitizedTextBlock));
+			for (String[] parsedLine : parsedLines) {
+				Preconditions.notNull(parsedLine,
+					() -> "Line at index " + index.get() + " contains invalid CSV: \"\"\"" + textBlock + "\"\"\"");
+				processNullValues(parsedLine);
+				index.getAndIncrement();
 			}
+			return parsedLines;
+		}
+		catch (Throwable throwable) {
+			throw handleCsvException(throwable, this.annotation);
+		}
+	}
+
+	private String[] parseLine(long index, String line) {
+		try {
+			String[] parsedLine = this.csvParser.parseLine(line + LINE_SEPARATOR);
+			Preconditions.notNull(parsedLine,
+				() -> "Line at index " + index + " contains invalid CSV: \"" + line + "\"");
+			processNullValues(parsedLine);
+			return parsedLine;
 		}
 		catch (Throwable throwable) {
-			handleCsvException(throwable, this.annotation);
+			throw handleCsvException(throwable, this.annotation);
 		}
-		Preconditions.notNull(parsedLine, () -> "Line at index " + index + " contains invalid CSV: \"" + line + "\"");
-		return parsedLine;
 	}
 
-	static void handleCsvException(Throwable throwable, Annotation annotation) {
+	private void processNullValues(String[] parsedLine) {
+		if (!this.nullValues.isEmpty()) {
+			for (int i = 0; i < parsedLine.length; i++) {
+				if (this.nullValues.contains(parsedLine[i])) {
+					parsedLine[i] = null;
+				}
+			}
+		}
+	}
+
+	/**
+	 * @return this method always throws an exception and therefore never
+	 * returns anything; the return type is merely present to allow this
+	 * method to be supplied as the operand in a {@code throw} statement
+	 */
+	static RuntimeException handleCsvException(Throwable throwable, Annotation annotation) {
 		UnrecoverableExceptions.rethrowIfUnrecoverable(throwable);
 		if (throwable instanceof PreconditionViolationException) {
 			throw (PreconditionViolationException) throwable;
 		}
 		throw new CsvParsingException("Failed to parse CSV input configured via " + annotation, throwable);
 	}
 
+	private static String sanitizeTextBlock(String textBlock) {
+		// Split and remove trailing blank lines.
+		String[] lines = NEW_LINE_REGEX.split(textBlock, 0);
+		// Reassemble the lines.
+		textBlock = String.join(LINE_SEPARATOR, lines);
```

Comment:
```suggestion
		// Strip all trailing whitespace at the end of the text block.
		textBlock = textBlock.replaceAll("\\s+?$", "");
```

FWIW, I considered this approach, and the tests still pass, but I'm not convinced that it's as robust as the implementation in this PR.

Thoughts?